### PR TITLE
ENH: Port cKDTree query methods to C++, allow pickling on Python 2, and add support for periodic queries

### PR DESCRIPTION
**Pull request by sturlamolden**
New WIP PR after rebase disaster.

### CODE REVIEW

File: scipy/spatial/ckdtree/src/ckdtree_count_neighbors.cxx [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+#include <ios>
+
+#define CKDTREE_METHODS_IMPL
+#include "ckdtree_cpp_decl.h"
+#include "ckdtree_cpp_methods.h"
+#include "ckdtree_cpp_exc.h"
+#include "ckdtree_cpp_rectangle.h"
+
+static void
+__count_neighbors_traverse(const ckdtree      *self,
```
Comment (ewmoore) [inline scipy/spatial/ckdtree/src/ckdtree_count_neighbors.cxx]: Aren't all double underscore names reserved in C++?

Comment (pv) [inline scipy/spatial/ckdtree/src/ckdtree_count_neighbors.cxx]: Does this help with the performance regressions?
https://pv.github.io/scipy-bench/#regressions?sort=3&dir=desc

Comment (sturlamolden) [inline scipy/spatial/ckdtree/src/ckdtree_count_neighbors.cxx]: The C++ compilers do not complain about the double underscores, but I'll check. They are in the Cython code, but it is trivial to remove them.

Comment (sturlamolden) [inline scipy/spatial/ckdtree/src/ckdtree_count_neighbors.cxx]: Double underscores are gone. They serve no purpose anyway in the C++ code.

**Comment by sturlamolden** [comment]
```query_ball_point``` is now ported to C++. With some minor modifications it should also be possible to multithread this function.

**Comment by sturlamolden** [comment]
```count_neighbors```has a test failure according to Travis, returns 34 instead of 33.

**Comment by sturlamolden** [comment]
```query_ball_tree```is ported to C++, ```sparse_distance_matrix```only remaining method to port.

```query_ball_point``` is multithreaded similarly to ```query```.

**Comment by sturlamolden** [comment]
All query methods now in C++

Added formerly requested option of returning other containers than ```dok_matrix``` from ```sparse_distance_matrix```. Test case needed for these.

**Comment by sturlamolden** [comment]
Remaining issues:
- testing on Windows
- code review, final bug fixing

**Comment by sturlamolden** [comment]
Travis is happy with this, it seems.

**Comment by rgommers** [comment]
Copying @pv's question that didn't get answered yet: does this help with the performance regressions? https://pv.github.io/scipy-bench/#regressions?sort=3&dir=desc ?

I did some browsing through all the ``cKDTree`` benchmarks and while we get a major improvement for 0.16.x there are a few things that got a lot slower. Would be good to copy benchmark results for this PR in a comment here before merging.

**Comment by sturlamolden** [comment]
I can do a systematic benchmark of cKDTree in 0.15.x, 0.16.x and this PR. That is clearly needed before this is ready to merge. I tried to read the benchmarks from @pv , but I am not quite sure what they measure and how to read them.

0.16.x provides two ways of constructing the kd-tree: sliding midpoint as before and median as in scikit-learn. I would not be surprised if the new method could give performance boost or regression depending on the query method. Since the median method makes kd-tree construction slower (but queries can be faster), the net effect can also depend on how much the kd-tree construction account for the whole run-time (e.g. it matters relatively more for smaller data sets than for large). The effect on query time will also depend on the distribution of the data, though @jakevdp thinks the median method is only inferior in contrived situations. The effect on run-time from the changes here probably depends on the hardware as well, and maybe also the compiler.

Another thing to note is that speed is not everything. This PR also allows the GIL to be released while doing the queries. In 0.15.x the GIL is never released; in 0.16.x only ```.query```releases the GIL; here all query methods release the GIL. This is rather important because id-tree queries can be lengthy and will lock-up the interpreter while they run with the GIL locked. Even if we do not multithread the queries, concurrent i/o will suffer, GUIs will not respond, and real-time graphics might not update in time. kd-trees are often used for computer graphics (e.g. in games and simulations), so this is rather important.

**Comment by sturlamolden** [comment]
I should probably set up a repo for the benchmarks.

**Comment by sturlamolden** [comment]
https://github.com/sturlamolden/cKDTree-bench

(No benchmarks yet, but it can be used to build ```cKDTree``` from 0.15.x, 0.16.x and this PR for correct comparisons.)

**Comment by sturlamolden** [comment]
Benchmarks:
http://nbviewer.ipython.org/github/sturlamolden/cKDTree-bench/blob/master/benchmark-i7.ipynb

**Comment by sturlamolden** [comment]
The only performance regression is in the kd-tree creation. But that is on a time scale three orders of magnitude smaller than the duration of the query methods.

**Comment by rgommers** [comment]
The existing benchmarks are here: ``benchmarks/benchmarks/spatial.py``. It looks like the benchmark results do depend on the shape of the input data.

The rendered benchmarks are at https://pv.github.io/scipy-bench/# (look for ``spatial``). They show performance as function of time, from 2009 to now. As an example of a regression, see https://pv.github.io/scipy-bench/#spatial.Neighbors.time_sparse_distance_matrix. On the left sidebar, deselect ``KDTree`` and select ``cKDTree`` under ``class``. Then you'll see a graph with a jump up in 2013 (the Cython rewrite) and other jump up (click and drag a zoom window onto the plot to zoom in) from a couple of weeks ago.

Now for a particular benchmark it can be the case that that benchmark is not realistic and can therefore be ignored or should be adapted. The ``query``, ``query_pairs`` and ``query_ball_points`` ones look OK. ``build`` becoming slower might be expected. ``sparse_distance_matrix`` becoming slower I'm not sure about.

Note that you can run the benchmarks locally with ``$ python runtests.py --bench -s spatial``

**Comment by rgommers** [comment]
The output you'll get is:

    $ python runtests.py --bench -s spatial
    Building, see build.log...
        ... build in progress
        ... build in progress
        ... build in progress
        ... build in progress
        ... build in progress
        ... build in progress
        ... build in progress
        ... build in progress
    Build OK
    Running benchmarks for Scipy version 0.17.0.dev0+7b893ae at /home/rgommers/Code/bldscipy/build/testenv/lib/python2.7/site-packages/scipy/__init__.pyc
    · Discovering benchmarks
    · Running 6 total benchmarks (1 commits * 1 environments * 6 benchmarks)
    [  0.00%] ·· Building for /usr/bin/python
    [  0.00%] ·· Benchmarking /usr/bin/python
    [ 16.67%] ··· Running spatial.Build.time_build                                                       ok
    [ 16.67%] ···· 
                   =================== ========= =========
                   --                         class       
                   ------------------- -------------------
                        (m, n, r)        KDTree   cKDTree 
                   =================== ========= =========
                     (3, 10000, 1000)   33.08ms    3.46ms 
                     (8, 10000, 1000)   64.40ms    4.73ms 
                    (16, 10000, 1000)   37.49ms    6.80ms 
                   =================== ========= =========

    [ 33.33%] ··· Running spatial.Neighbors.time_count_neighbors                                         ok
    [ 33.33%] ···· 
                   ================== ============== ========== =========
                   --                                       class        
                   --------------------------------- --------------------
                      (m, n1, n2)      probe radius    KDTree    cKDTree 
                   ================== ============== ========== =========
                    (3, 1000, 1000)        0.2        410.96ms    2.32ms 
                    (3, 1000, 1000)        0.5        698.07ms    3.90ms 
                    (8, 1000, 1000)        0.2         1.71s      8.46ms 
                    (8, 1000, 1000)        0.5         2.11s     11.01ms 
                    (16, 1000, 1000)       0.2         2.50s     14.63ms 
                    (16, 1000, 1000)       0.5         2.71s     17.44ms 
                   ================== ============== ========== =========

    [ 50.00%] ··· Running spatial.Neighbors.time_sparse_distance_matrix                                  ok
    [ 50.00%] ···· 
                   ================== ============== ======== =========
                   --                                      class       
                   --------------------------------- ------------------
                      (m, n1, n2)      probe radius   KDTree   cKDTree 
                   ================== ============== ======== =========
                    (3, 1000, 1000)        0.2        1.41s     1.97ms 
                    (3, 1000, 1000)        0.5        2.53s     6.99ms 
                    (8, 1000, 1000)        0.2        5.39s     6.83ms 
                    (8, 1000, 1000)        0.5        6.59s     9.42ms 
                    (16, 1000, 1000)       0.2        9.42s    14.19ms 
                    (16, 1000, 1000)       0.5        10.68s   17.21ms 
                   ================== ============== ======== =========

    [ 66.67%] ··· Running spatial.Query.time_query                                                       ok
    [ 66.67%] ···· 
                   =================== ========== ========== ==============
                   --                                 class                
                   ------------------- ------------------------------------
                        (m, n, r)        KDTree    cKDTree    cKDTree_flat 
                   =================== ========== ========== ==============
                     (3, 10000, 1000)   247.76ms    2.73ms      72.60ms    
                     (8, 10000, 1000)    4.26s     22.65ms      105.84ms   
                    (16, 10000, 1000)    27.65s    235.68ms     166.01ms   
                   =================== ========== ========== ==============

    [ 83.33%] ··· Running spatial.Radius.time_query_ball_point                                           ok
    [ 83.33%] ···· 
                   ================== ============== ======== ========= ==============
                   --                                              class              
                   --------------------------------- ---------------------------------
                       (m, n, r)       probe radius   KDTree   cKDTree   cKDTree_flat 
                   ================== ============== ======== ========= ==============
                    (3, 10000, 1000)       0.2        2.08s     8.49ms     77.31ms    
                    (3, 10000, 1000)       0.5        6.04s    26.86ms     83.03ms    
                   ================== ============== ======== ========= ==============

    [100.00%] ··· Running spatial.Radius.time_query_pairs                                                ok
    [100.00%] ···· 
                   ================ ============== ========== ========== ==============
                   --                                             class                
                   ------------------------------- ------------------------------------
                      (m, n, r)      probe radius    KDTree    cKDTree    cKDTree_flat 
                   ================ ============== ========== ========== ==============
                    (3, 1000, 30)        0.2        388.76ms   977.07μs      3.75ms    
                    (3, 1000, 30)        0.5        506.58ms    1.63ms       4.29ms    
                    (8, 1000, 30)        0.2         2.51s      3.86ms       5.72ms    
                    (8, 1000, 30)        0.5         2.46s      4.72ms       5.67ms    
                    (16, 1000, 30)       0.2         2.63s      7.39ms       8.66ms    
                    (16, 1000, 30)       0.5         2.62s      8.11ms       8.71ms    
                   ================ ============== ========== ========== ==============

**Comment by sturlamolden** [comment]
```cKDTree``` did not have ```sparse_distance_matrix``` prior to the Cython rewrite. Clearly that could not have caused a regression :)

**Comment by sturlamolden** [comment]
In my tests I used the dummy data generator from scikit-learn to get a more realistic dataset than just random numbers.

**Comment by pv** [comment]
sparse_distance_matrix has been in cKDTree since 882820eb1f923d4d73d in
2012. Obviously, the regression is after that.

**Comment by pv** [comment]
The first regression in the distance matrix is probably due to fixing the implementation to give correct result, probably a9565c0a711

The kdtree benchmarks currently under benchmarks/ are those that were previously under scipy/spatial/benchmarks --- I have no idea whether they are sensible or not. If you think they are not sensible, they should be replaced.

**Comment by pv** [comment]
The more recent performance regression in sparse_distance_matrix benchmarks is bea41ad6, as indicated in the plots.

**Comment by sturlamolden** [comment]
Did you look at the timings in the ipython notebook? Here you can see the runtime of ```sparse_distance_matrix``` redused from 4.07 s to 3.12 s from SciPy 0.15.x to 0.16.x. Yet your benchmark show a performance regression. I am not quite sure what is going on here.

**Comment by sturlamolden** [comment]
What I mean by "realistic dataset" is that the data has some structure so that the n-dimensional sample space is not uniformly populated. This is at least the common situation is classification problems.

**Comment by pv** [comment]
Note that it's not really "my" benchmark --- these are the old benchmarks written by the original ckdtree authors. You can try running them on your machine and to see if you can reproduce the results.

**Comment by sturlamolden** [comment]
Ok, sorry, bad wording :)

**Comment by pv** [comment]
Well, I'm just trying to say that I have even less idea what's going on than you do :)

**Comment by sturlamolden** [comment]
I created the repo https://github.com/sturlamolden/cKDTree-bench to make sure everything except the ```cKDTree``` code is kept constant. In SciPy there are more degrees of freedom than just the Cython and C++ code.

**Comment by pv** [comment]
Note that the same thing can be achieved by benchmarking each commit separately, so that observed performance changes are localized to a single merge commit, so that this does not affect the above conclusions.

**Comment by sturlamolden** [comment]
@rgommers I see a spike at ```ba4fbd5c``` and another at ```dc313cb5```, but they are not related to ```cKDTree``` and then it goes down again. Noise? It seems to me the run-time of ```sparse_distance_matrix``` has been rather constant. 

https://dl.dropboxusercontent.com/u/12464039/Screen%20Shot%202015-05-22%20at%2016.48.42.png

**Comment by pv** [comment]
@sturlamolden: Note that the cKDTree results in that picture are all close to zero, so you can't see the changes. If you turn off the KDTree lines (from the left-hand panel), you can probably see better what is changing.

**Comment by sturlamolden** [comment]
@pv Thanks. Now I see it:
https://dl.dropboxusercontent.com/u/12464039/Screen%20Shot%202015-05-22%20at%2017.09.14.png

Query improved though:
https://dl.dropboxusercontent.com/u/12464039/Screen%20Shot%202015-05-22%20at%2017.01.13.png

**Comment by sturlamolden** [comment]
I have an idea what this might be...

**Comment by sturlamolden** [comment]
Ok, so replicating exactly the benchmark ```(16,1000,1000),0.5``` it goes up from ```16.1 ms``` in 0.15.x to ```18.6 ms``` in 0.16.x. On the other hand with ```(8,1000,1000),0.5``` it improves from ```16.3 ms``` in 0.15.x to ```8.71 ms``` in 0.16.x. It seems the effect on runtime in ```sparse_distance_matrix``` depends on the data.

**Comment by sturlamolden** [comment]
It seems to be CPU and compiler dependent as well. 

I will add prefetching to ```sparse_distance_matrix``` as well. It seems AMD processors benefit more from explicit prefetching than Intel processors.

**Comment by sturlamolden** [comment]
I have been thinking about how to multi-thread the query methods that compare two kd-trees. That is, ```query_ball_tree```, ```query_pairs```, ```count_neighbors```, and ```sparse_distance_matrix```. The easiest solution I can see is to construct a new set of kd-trees from one of the trees. It will increase the startup overhead, but kd-tree contruction is very cheap compared to kd-tree queries. That should allow these query methods to run in parallel without any contention between the threads.

**Comment by aarchiba** [comment]
Aren't the kD-tree objects read-only during query operations? If so, multiple queries should not interfere with each other (at all, if you can give the CPU the right cache-coherency promises, but that might require non-portable jiggery-pokery). Failing that, an operation to duplicate kD-trees (sub-kD-trees?) should be cheaper than building new ones.

**Comment by sturlamolden** [comment]
Yes, they are declared ```const ckdtree *``` in the C++ code. But I am not sure how to split up the work load without duplicating one of them.

**Comment by aarchiba** [comment]
I see now what you mean about splitting up the kD-trees. It does seem like the simplest method and the one requiring the least interaction between threads. Could effort be saved by splitting up an existing kD-tree into its natural sub-trees? Of course that makes it likely that the query on one branch will finish long before the query on another branch. To even out the work I guess random subsets for each thread are safest.

Or if thread interaction isn't too expensive, the priority queue (heap) used for tree traversal could be treated as a task pool: each thread grabs the top-priority remaining item, does the work, and queues whichever further nodes need traversal, then goes looking for another job. Very probably the constant locking would make this hopelessly slow.

**Comment by sturlamolden** [comment]
I also think I should change some of the long function names that Pat made back to what you used in the Python kd-tree code. E.g. ```query_ball_tree_traverse_checking``` will just be ```traverse_checking```. I think those long names had some justification in the Cython code. But that is clearly not needed here.

**Comment by sturlamolden** [comment]
The two builds methods in 0.16.x is now merged into one method. 

Having the build method in C++ means we can release the GIL while the ```cKDTree``` is constructed. This should allow us to multi-thread creation of sub-trees when doing parallel queries with the methods that compares two  ```cKDTree``` instances.

The Cython code is now a plain wrapper that exposes a an API to Python. All algorithmic code is factored out into multiple .cxx and .h files. This makes it more readable than having one monolithic Cython file.

**Comment by sturlamolden** [comment]
Using the test data set that SciPy has in its benchmark I am seeing timings like this (single-threaded code). 

```count_neighbours``` and ```query_ball_point``` have excellent speed-ups. 

```query_ball_tree```, ```query_pairs``` and ```sparse_distance_matrix``` is not much better than in 0.16.x, but they now release the GIL and can be multithreaded. 

```
count_neighbours
0.15.x  138 ms
0.16.x   50.5 ms
#4849   17.1 ms

query_ball_tree
0.15.x   36.6 ms
0.16.x   18.4 ms
#4849   16.4 ms

query_ball_point
0.15.x   18.9 ms
0.16.x   18.1 ms
#4849    6.0 ms

query_pairs
0.15.x   8.4 ms
0.16.x   9.5 ms
#4849   8.9 ms

sparse_distance_matrix
0.15.x   15.6 ms
0.16.x   17.7 ms
#4849   16.8 ms
```

**Comment by sturlamolden** [comment]
Travis timed out (50 min) the full coverage test again.

**Comment by sturlamolden** [comment]
figure_1

```cKDTree.query_ball_point``` has this note

```
Notes
        -----
        If you have many points whose neighbors you want to find, you may save
        substantial amounts of time by putting them in a cKDTree and using
        query_ball_tree.
```

Clearly that is no longer the case, ```query_ball_tree``` is only faster for smaller data sets. Is it ok if I remove it? @aarchiba

**Comment by sturlamolden** [comment]
figure_1

The upper was for 4 dimensions, if I increase to 8 dimensions I get this, in which case ```query_ball_tree``` is actually the slower.

**Comment by sturlamolden** [comment]
Given that ```query_ball_tree```is clearly inferior, should we just implement it on top of ```query_ball_point```? Or change the note to advice the user to use ```query_ball_point```instead? @rgommer @aarchiba

**Comment by aarchiba** [comment]
I don't understand why query_ball_tree isn't faster than the single-threaded query_ball_point. Substantial pruning should be possible. Perhaps it requires clever two-tree-aware traversal algorithms? In any case, as you say, the two-tree version isn't any faster. I'd be inclined to simply switch its implementation to query_ball_point, though, in case someone comes along and writes an efficient two-tree traversal algorithm that speeds things up later. A note saying that it isn't any faster so don't bother building a tree if you don't already have one might be good too.

**Comment by sturlamolden** [comment]
There is also a strange inconsistency in the API, which is that 
```kdtree1.query_ball_point(kdtree2.data,r)```
corresponds to
```kdtree2.query_ball_tree(kdtree1,r)```
So they basically transpose the output. 

Another thing is that all the query methods apart from ```query``` is actually doing the same. So there is a lot of code duplication.

**Comment by jakevdp** [comment]
I'd be +1 on using ``query_radius``. The ``query_ball_tree`` name has always seemed overly confusing, especially given that "Ball Tree" is the name of a common data structure closely related to the KD Tree.

**Comment by sturlamolden** [comment]
By the way, I have also noted in scikit-learn that ```query``` and ```query_radius``` have similar inconsistent output. One returns an array of distances and an array of indices, the other returns an array of ```(index, distance)``` tuples. I am not sure why. The good thing here is that I can make the APIs in SciPy consistent without breaking anyones code ;-)

As for ball-trees, we need those in SciPy, and vp-trees would be cool too. It is not difficult to implement those with minor changes to the kd-tree code. But, that will have to wait for my next WIP PR :)

```query_ball_tree``` should probably have a statement clarifying that it has nothing to do with a ball tree.

**Comment by jakevdp** [comment]
> the other returns an array of (index, distance) tuples.

I don't think this is true... though you're right that it's inconsistent. One returns a tuple of ``(dist, ind)`` arrays, while the other returns a tuple of ``(ind, dist)`` arrays.

**Comment by sturlamolden** [comment]
That is probably the same reason why ```query_ball_point``` and ```query_ball_tree``` are inconsistent in SciPy. The choise of output is arbitrary, and we can easily transform from one to the other. But I guess nobody noticed when doing code review (if there was a review).

**Comment by sturlamolden** [comment]
Please notice that the C++ code have carried over the logic error reported in #5077 It needs to be corrected before it can be merged.

**Comment by sturlamolden** [comment]
It seems I'we created a rebase disaster again.

**Comment by rgommers** [comment]
@sturlamolden at least you're consistent:) 

You've lost a whole bunch of commits it looks like. These to be exact:

    * c2e77d6 - (HEAD, upstream/pr/4890, pr/4890) MAINT: update kwarg when calling sparse_distance_matrix (6
    * ed486a8 - MAINT: refactor query_pairs and sparse_distance_matrix (6 weeks ago)
    * ddaf97a - BUG: define NPY_UNLIKELY for building on NumPy 1.6 (7 weeks ago)
    * 8500609 - MAINT: C++ code readability improvements (7 weeks ago)
    * 22bcc9b - MAINT: remove reduntant file (7 weeks ago)
    * 889ae89 - ENH: move build method to C++ (7 weeks ago)
    * 09257ce - MAINT: stylistic makeover (8 weeks ago)
    * 32a1b92 - ENH: prefetching in sparse_distance_matrix (9 weeks ago)

I still have them, so let me know if you need help recovering them.

**Comment by sturlamolden** [comment]
Thankfully I have all the code on my computer, since I keep a "repo" managed by Apple time machine. It seems to better fit my skills 😂 

I am not sure what I did, I just wanted to rebase on the current master because it was lagging behind. Obviously I did something else as well...😡

The Git repo seems to be fubar, though. Every time I try to do something with Git, bad things happens. Also, I might have lost those commits, but the code changes they contained are still there. But the changes might have been reintroduced in a later commit, I am not sure. Also I made Travis complain, but I that might be due to something else.

I have no idea how to recover those commits, or even if I should try.

**Comment by rgommers** [comment]
Here is a rebased branch with what look like sane commits: https://github.com/rgommers/scipy/tree/pr-4890-fixup. You can take it over by:

    git remote add rgommers https://github.com/rgommers/scipy.git
    git fetch rgommers
    git checkout -b pr-4890-fixup rgommers/pr-4890-fixup
    # check that the two branches contain the same end result, and if so
    git br -D cpp_ckdtree
    git checkout -b cpp_ckdtree pr-4890-fixup  # now you have the same name branch with the right commits

**Comment by sturlamolden** [comment]
Thanks, you are almost being too helpful 😊

I have to learn Git better and stop creating this extra work for you.

**Comment by rgommers** [comment]
Well, I don't know the first thing about what is idiomatic in C++, so we're trading in kind:)

**Comment by sturlamolden** [comment]
This is more like C with classes than C++ ;)

**Comment by sturlamolden** [comment]
Pulled in the code from @rainwoodman which implements periodic spaces (for cosmology simulations).

I am not sure if the Travis build will pass. But that will be fixed later.

**Comment by sturlamolden** [comment]
As expected there are some minor issues here which I will have to fix:
- kwarg misspelling in the test suite
- unknown segfault on Python 3.3 and 3.4, ok on 3.2 and 2.7
- an error building the documentation

**Comment by ev-br** [comment]
`linear_sum_assignment` refguide-check failure is unrelated. I'll fix it now in https://github.com/scipy/scipy/pull/5247 so that you guys can focus on real stuff :-).

**Comment by rainwoodman** [comment]
I couldn't reproduce the segfault with a plain python3.4.1. So it could be more complicated. How to set up a Travis like environment?

**Comment by rainwoodman** [comment]
Looks like we need a rebase?

**Comment by sturlamolden** [comment]
Yes, we need to rebase and also to rewrite some of the code. ```infinity``` has been taken out as a global variable to prevent a linker error on SunOS. So basically we have to change all our ```np.inf``` handling to  use ```NPY_INFITINITY``` from ```numpy/npy_math.h```. It is rather pervasive.  We also need to incorporate the ```ckdtree_isininity``` function now in master, because we cannot use ```npy_isininity``` because of performance and because of a build error. There were some changes to the Cython code as well. So there is a maintenance job todo.

**Comment by sturlamolden** [comment]
@rainwoodman I was referring to this: https://github.com/scipy/scipy/pull/5314

**Comment by rainwoodman** [comment]
So the idea is to replace every comparison with infinity to ckdtree_isinfinity.

But how to resolve the conflicts? -- rebase+resolve merge-conflicts in each commit, or add more commits?

**Comment by sturlamolden** [comment]
It needs more commits. E.g. we have to get rid of ```infinity``` and ```::infinity``` as local and global variables and replace with ```NPY_INFINITY```. This also includes changing how the distance functions are called. This PR got very out of synch with master after the fix for SunOS was included. I am an idiot with Git so I am not sure how to properly solve the conflicts. 

Another thing is should this PR be bumped to 0.18 if 0.17.0 will be branched soon?

**Comment by rainwoodman** [comment]
Experience I recently gained suggests we shall probably

1 merge the scipy master,
2 either use 'checkout --theirs' or 'checkout --ours', to keep our files
3 then fix the in consistent parts.

Does this plan make sense?

**Comment by rgommers** [comment]
> Another thing is should this PR be bumped to 0.18 if 0.17.0 will be branched soon?

Don't worry about that. We don't have a planned date for branching yet. And stuff like ``. versionadded:: 0.17.0`` can easily be changed right before the merge of a PR.

**Comment by rgommers** [comment]
@rainwoodman I think that that plan makes sense. I gave it a quick try, but this PR isn't easily rebase-able. So merging with master it is. I don't know too much about ``--theirs/ours``, but in this case it seems that merging your branch into an up-to-date branch off master is much cleaner than the other way around. So:

    $ git co -b cpp_ckdtree2 master
    $ git merge cpp_ckdtree

gives the expected output instead of many many new files:

    On branch cpp_ckdtree2
    You have unmerged paths.
      (fix conflicts and run "git commit")

    Changes to be committed:

            modified:   scipy/spatial/bento.info
            new file:   scipy/spatial/ckdtree.pyx
            new file:   scipy/spatial/ckdtree/src/build.cxx
            deleted:    scipy/spatial/ckdtree/src/ckdtree_cpp_methods.h
            renamed:    scipy/spatial/ckdtree/src/ckdtree_cpp_decl.h -> scipy/spatial/ckdtree/src/ckdtree_decl.h
            new file:   scipy/spatial/ckdtree/src/ckdtree_methods.h
            new file:   scipy/spatial/ckdtree/src/converter.h
            new file:   scipy/spatial/ckdtree/src/coo_entries.h
            new file:   scipy/spatial/ckdtree/src/count_neighbors.cxx
            renamed:    scipy/spatial/ckdtree/src/ckdtree_cpp_exc.cxx -> scipy/spatial/ckdtree/src/cpp_exc.cxx
            renamed:    scipy/spatial/ckdtree/src/ckdtree_cpp_exc.h -> scipy/spatial/ckdtree/src/cpp_exc.h
            renamed:    scipy/spatial/ckdtree/src/ckdtree_cpp_utils.h -> scipy/spatial/ckdtree/src/cpp_utils.h
            new file:   scipy/spatial/ckdtree/src/distance.h
            new file:   scipy/spatial/ckdtree/src/distance_box.h
            renamed:    scipy/spatial/ckdtree/src/ckdtree_globals.cxx -> scipy/spatial/ckdtree/src/globals.cxx
            new file:   scipy/spatial/ckdtree/src/ordered_pair.h
            new file:   scipy/spatial/ckdtree/src/partial_sort.h
            new file:   scipy/spatial/ckdtree/src/query.cxx
            new file:   scipy/spatial/ckdtree/src/query_ball_point.cxx
            new file:   scipy/spatial/ckdtree/src/query_ball_tree.cxx
            new file:   scipy/spatial/ckdtree/src/query_pairs.cxx
            new file:   scipy/spatial/ckdtree/src/rectangle.h
            new file:   scipy/spatial/ckdtree/src/sparse_distances.cxx
            modified:   scipy/spatial/setup.py

    Unmerged paths:
      (use "git add/rm <file>..." as appropriate to mark resolution)

            deleted by them:    scipy/spatial/ckdtree/ckdtree.pyx
            deleted by them:    scipy/spatial/ckdtree/src/ckdtree_query.cxx
            both modified:      scipy/spatial/ckdtree/src/query_methods.h
            both modified:      scipy/spatial/tests/test_kdtree.py

So only fix the merge commits here and commit.

**Comment by sturlamolden** [comment]
```bash
Sturla-laptop:scipy sturla$ git checkout -b cpp_ckdtree2 scipy-master
Switched to a new branch 'cpp_ckdtree2'
Sturla-laptop:scipy sturla$ git merge cpp_ckdtree
Auto-merging scipy/spatial/tests/test_kdtree.py
CONFLICT (content): Merge conflict in scipy/spatial/tests/test_kdtree.py
Auto-merging scipy/spatial/setup.py
Auto-merging scipy/spatial/ckdtree/src/query_bashmethods.h
CONFLICT (content): Merge conflict in scipy/spatial/ckdtree/src/query_methods.h
Auto-merging scipy/spatial/ckdtree/src/cpp_utils.h
CONFLICT (modify/delete): scipy/spatial/ckdtree/src/ckdtree_query.cxx deleted in cpp_ckdtree and modified in HEAD. Version HEAD of scipy/spatial/ckdtree/src/ckdtree_query.cxx left in tree.
Auto-merging scipy/spatial/ckdtree/src/ckdtree_decl.h
CONFLICT (modify/delete): scipy/spatial/ckdtree/ckdtree.pyx deleted in cpp_ckdtree and modified in HEAD. Version HEAD of scipy/spatial/ckdtree/ckdtree.pyx left in tree.
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by sturlamolden** [comment]
It remains to be seen if I did this right...

**Comment by rgommers** [comment]
looks right to me

**Comment by rainwoodman** [comment]
The steps are clear. I will take a stab this weekend. Any more potential issues than 
the infinite fixes in 
https://github.com/scipy/scipy/pull/4890#issuecomment-151752292

I noticed a recent issue regarding npy_float64 -- is it relevant?

**Comment by sturlamolden** [comment]
I have done the merge and I will push that soon. I just have to fix a couple of merge conflicts.

If you mean the ```npy_float64``` issue reported on the NumPy mailing list it is the https://github.com/scipy/scipy/pull/4890#issuecomment-151752292 problem. It seems that ```::infinite``` causes a name collision on SunOS and Cygwin.

**Comment by rainwoodman** [comment]
Yes. Thanks.

**Comment by sturlamolden** [comment]
Stupid question, but how to I get the contents from my local branch ```cpp_ckdtree2``` into the remote beanch ```origin/cpp_ckdtree```?

**Comment by pv** [comment]
git push origin cpp_ckdtree2:cpp_ckdtree
And with "push -f" in case of rebasing etc.

**Comment by sturlamolden** [comment]
Thanks @pv, I'll try to remember that 

However I had already done this:
```
sturla$ git branch -d cpp_ckdtree
Deleted branch cpp_ckdtree (was 475a71c).
sturla$ git branch cpp_ckdtree
sturla$ git checkout cpp_ckdtree
Switched to branch 'cpp_ckdtree'
sturla$ git push -f origin
```

**Comment by sturlamolden** [comment]
I am not expecting the build to succeed, but now the merge conflicts are taken care of. The next step is to fix the build issue on SunOS and Cygwin due to ```::infinite```.

**Comment by sturlamolden** [comment]
But on the positive side, we are up to date with scipy/master again :-)

**Comment by sturlamolden** [comment]
@rainwoodman  What is the purpose of the lambda returning None here?

```Python
def test_ckdtree_box():
    # check ckdtree periodic boundary
    #
    self = lambda x : None
    self.n = 2000
    self.m = 2
    self.k = 3
    np.random.seed(1234)
    data = np.random.uniform(size=(self.n, self.m))
    kdtree = cKDTree(data, leafsize=1, boxsize=1.0)

    # use the standard python KDTree for the simulated periodic box
    kdtree2 = cKDTree(data, leafsize=1)
#    data = data[1:2]

    dd, ii = kdtree.query(data, self.k)

    dd1, ii1 = kdtree.query(data + 1.0, self.k)
    assert_almost_equal(dd, dd1)
    assert_equal(ii, ii1)
    
    dd1, ii1 = kdtree.query(data - 1.0, self.k)
    assert_almost_equal(dd, dd1)
    assert_equal(ii, ii1)

    dd2, ii2 = simulate_periodic_box(kdtree2, data, self.k, boxsize=1.0)
    assert_almost_equal(dd, dd2)
    assert_equal(ii, ii2)
```

**Comment by sturlamolden** [comment]
As expected, all the builds are failing because the symbol ```::infinity``` was removed in scipy/master and is now missing.

**Comment by sturlamolden** [comment]
Bugger, I forgot about using prefixes on the commit messages :-(

**Comment by sturlamolden** [comment]
Great, what do you think @rainwoodman ? Are we happy now? More that needs to be done? Fix my last commit messages perhaps (I don't know how), but apart from that, does this look good?

**Comment by sturlamolden** [comment]
@rgommers Should we put an 0.17 milestone on this? This PR is basically done, unless @rainwoodman has anything else to add. 

(It remains to verify that it builds on Windows. My old Windows PC just died of old age, but I am setting up a virtual box to run on my Mac.)

**Comment by sturlamolden** [comment]
Summary of what is in this PR:
- All query methods are in C++ and release the GIL
- Build method is in C++ and releases the GIL
- Support for periodic spaces for using cKDTree in cosmology simulations
- Pickling is supported on Python 2 as well
- ```sparse_distance_matrix``` can return other containers than ```dok_matrix``` which prevents truncation to 0 of distances between excluded pairs.
- Many conditional branches are replaced with C++ templates to ensure the fastest execution path
- ```query_ball_point``` can optionally be multi-threaded

**Comment by rainwoodman** [comment]
Sure. I am happy with this. Thanks!

**Comment by sturlamolden** [comment]
WIP tag is removed.

**Comment by rainwoodman** [comment]
We will likely write a code for 2-point galaxy clustering analysis in cosmology, with an extension to the original Gray and Moore 2000 to the anisotropic case. 

This code will be based on the current ckdtree code in this branch, but would add these features to ckdtree:

- weighted input data, multiple weights shall be supported. The tree nodes will be augmented to include the sum of weights for proper pruning.
``` cKDTree(data, leafsize, boxsize, weight) ```

- An anisotropic count_neighbors method with pruning, that returns number of pairs as a function of r and z/r, where z is the distance along a selected axis:
``` count_neighbors_2d(self, other, r, mu, p) ```

From my eyes these two features are quite doable. What do your think? @sturlamolden  @rgommers ?

We will be testing the full program (which would just blend some parallelism and IO interface with the modified ckdtree) against the standard code used in SDSS galaxy clustering analysis (https://github.com/bareid/xi), and fix bugs along the way. 

I am guessing the time-line for a new ckdtree PR will be the next 2 to 3 months, probably not soon enough to make it into next release?

**Comment by sturlamolden** [comment]
They are probably doable. I am not sure what the weights do, though, but now that we use templates it is easy to plug them in without producing a performance regression.

0.17 will probably be tagged and branched soon. It is that time of the year. Then 0.18 will be next summer, or something like that. :-)

**Comment by sturlamolden** [comment]
Can someone please tag this PR with 0.17 milestone?

**Comment by rgommers** [comment]
I've verified that this builds and passes all tests with MinGW (python 3.4, sse3 build). So in it goes. Thanks @sturlamolden and @rainwoodman!

**Comment by rgommers** [comment]
@rainwoodman a next release is never more than 6 months away, so no worries about the timing for the next PR.

**Comment by rgommers** [comment]
I'll send an update for the release notes.

**Pull request merged**